### PR TITLE
8298084: Memory leak in Method::build_profiling_method_data

### PR DESCRIPTION
**Pull request by coleenp**
This change fixes the MethodData leak by calling the destructor in both the release_C_heap_structures conditionally and by calling the MethodData destructor in the MetadataFactory::free_metadata method.
Thanks to @jcking for working on the patch and discussion.
Tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298084](https://bugs.openjdk.org/browse/JDK-8298084): Memory leak in Method::build_profiling_method_data

### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [e195cd85](https://git.openjdk.org/jdk20/pull/13/files/e195cd854e801ec4fd0b7fc5723e4296753d90be)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to [11fcdba0](https://git.openjdk.org/jdk20/pull/13/files/11fcdba0d70e4459fbecd88dd519d4acf73bce30)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Justin King](https://openjdk.org/census#jcking) (@jcking - Author)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Contributors
 * Justin King `<jcking@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk20 pull/13/head:pull/13` \
`$ git checkout pull/13`

Update a local copy of the PR: \
`$ git checkout pull/13` \
`$ git pull https://git.openjdk.org/jdk20 pull/13/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13`

View PR using the GUI difftool: \
`$ git pr show -t 13`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk20/pull/13.diff">https://git.openjdk.org/jdk20/pull/13.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/memory/metadataFactory.hpp [hunk 1]
```diff
@@ -65,10 +67,17 @@ class MetadataFactory : AllStatic {
...
+      // Call metadata's deallocate function which will deallocate fields and release_C_heap_structures
       assert(!md->on_stack(), "can't deallocate things on stack");
       assert(!md->is_shared(), "cannot deallocate if in shared spaces");
       md->deallocate_contents(loader_data);
+      // If requested, call the destructor. This is currently used for MethodData which has a member
+      // that needs to be destructed to release resources. Most Metadata derived classes have noop
+      // destructors and/or cleanup using deallocate_contents.
+      // T is a potentially const or volatile qualified pointer. Remove the pointer and any const
+      // or volatile so we can call the destructor of the type T points to.
+      using U = typename RemoveCV<typename RemovePointer<T>::type>::type;
```
Comment (kimbarrett) [inline src/hotspot/share/memory/metadataFactory.hpp]: Use `std::remove_pointer<T>` and `std::remove_cv<T>` from `<type_traits>` in
preference to our pre-C++11/14 workarounds. You can also avoid `typename` and
`::type` by using the "_t"-suffixed versions. So
```
using U = std::remove_cv_t<std::remove_pointer_t<T>>;
```
And "If requested" in the comment? That suggests some conditionalization, of
which there is none. Maybe something like "Most of the cleanup is done via
deallocate_contents, but some types require some additional cleanup via the
destructor."

**Review by kimbarrett — CHANGES_REQUESTED** [review]

File: src/hotspot/share/memory/metadataFactory.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/memory/metadataFactory.hpp]: I fixed this as suggested.  Thanks.

File: src/hotspot/share/memory/metadataFactory.hpp [hunk 2]
```diff
@@ -71,12 +70,12 @@ class MetadataFactory : AllStatic {
...
       assert(!md->is_shared(), "cannot deallocate if in shared spaces");
       md->deallocate_contents(loader_data);
-      // If requested, call the destructor. This is currently used for MethodData which has a member
+      // Call the destructor. This is currently used for MethodData which has a member
       // that needs to be destructed to release resources. Most Metadata derived classes have noop
       // destructors and/or cleanup using deallocate_contents.
       // T is a potentially const or volatile qualified pointer. Remove the pointer and any const
       // or volatile so we can call the destructor of the type T points to.
-      using U = typename RemoveCV<typename RemovePointer<T>::type>::type;
+      using U = std::remove_cv_t<std::remove_pointer_t<T>>;
```
Comment (kimbarrett) [inline src/hotspot/share/memory/metadataFactory.hpp]: Could get rid of `remove_pointer` by changing the function parameter from `T md` to `T* md`.  Apologies for not thinking of that earlier.

**Review by kimbarrett — APPROVED** [review]

File: src/hotspot/share/memory/metadataFactory.hpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/memory/metadataFactory.hpp]: The compiler seems not unhappy with that.

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by coleenp** [comment]
Thanks Erik and Kim.  @TobiHartmann and @jcking, do you want to review?

File: src/hotspot/share/memory/metadataFactory.hpp [hunk 3]
```diff
@@ -61,14 +62,21 @@ class MetadataFactory : AllStatic {
...
-      // Call metadata's deallocate function which will call deallocate fields
+      // Call metadata's deallocate function which will deallocate fields and release_C_heap_structures
       assert(!md->on_stack(), "can't deallocate things on stack");
       assert(!md->is_shared(), "cannot deallocate if in shared spaces");
       md->deallocate_contents(loader_data);
+      // Call the destructor. This is currently used for MethodData which has a member
+      // that needs to be destructed to release resources. Most Metadata derived classes have noop
+      // destructors and/or cleanup using deallocate_contents.
+      // T is a potentially const or volatile qualified pointer. Remove the pointer and any const
+      // or volatile so we can call the destructor of the type T points to.
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/metadataFactory.hpp]: Comment needs updating now you don't need to remove the pointer.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 4]
```diff
@@ -2665,12 +2665,14 @@ static void method_release_C_heap_structures(Method* m) {
...
-void InstanceKlass::release_C_heap_structures(bool release_constant_pool) {
+void InstanceKlass::release_C_heap_structures(bool release_sub_metadata) {
   // Clean up C heap
   Klass::release_C_heap_structures();
 
   // Deallocate and call destructors for MDO mutexes
-  methods_do(method_release_C_heap_structures);
+  if (release_sub_metadata) {
+    methods_do(method_release_C_heap_structures);
+  }
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp:2675]: I don't understand this aspect of the change. It seems unrelated to the missing destructor issue and seems to cause a significant change in behaviour as we will no longer call `methods_do` when called from `deallocate_contents` with the renamed `false` argument.

**Review by dholmes-ora — COMMENTED** [review]
Not sure about some of the incidental changes - see below.

Thanks.

File: src/hotspot/share/memory/metadataFactory.hpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/memory/metadataFactory.hpp]: Fixed.  Thanks for catching that.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.cpp:2675]: This boolean prevents us from calling the MethodData destructor twice and releasing MethodData C heap memory twice.  InstanceKlass::deallocate_contents calls release_C_heap_structures with it false, and then calls through the Metadata pointer hierarchy to MetadataFactory::free_metadata, which calls the MethodData destructor and MethodData::deallocate_contents which calls MethodData::release_C_heap_structures().   This is the delegation scheme for releasing metadata between InstanceKlass and ConstantPool.

InstanceKlass::release_C_heap_structures() is also called during class unloading and must walk the whole metadata, releasing C heap memory.  It's passed true. It doesn't go through deallocate_contents because the metadata is freed together at the same time with the metaspace that contains it.

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 5]
```diff
@@ -598,7 +598,7 @@ void InstanceKlass::deallocate_contents(ClassLoaderData* loader_data) {
   // Can't release the constant pool here because the constant pool can be
   // deallocated separately from the InstanceKlass for default methods and
   // redefine classes.
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp]: Does this comment make sense when the parameter doesn't actually control release of the CP?

File: src/hotspot/share/oops/instanceKlass.cpp [hunk 6]
```diff
@@ -2665,12 +2665,14 @@ static void method_release_C_heap_structures(Method* m) {
 }
 
 // Called also by InstanceKlass::deallocate_contents, with false for release_constant_pool.
```
Comment (dholmes-ora) [inline src/hotspot/share/oops/instanceKlass.cpp]: Comment needs updating.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/share/oops/instanceKlass.cpp:2675]: Thanks Coleen I see now how the code has been shuffled around:

Old code:
```
InstanceKlass::deallocate_contents
-> release_C_heap_structures(/* release_sub_metadata */ false);
  -> methods_do(method_release_C_heap_structures);
    -> m->release_C_heap_structures
      -> method_data()->~MethodData();
-> deallocate_methods(loader_data, methods());
  -> MetadataFactory::free_metadata(loader_data, method);
```
New code:
```
InstanceKlass::deallocate_contents
-> release_C_heap_structures(/* release_sub_metadata */ false);
-> deallocate_methods(loader_data, methods());
  -> MetadataFactory::free_metadata(loader_data, method);
    -> method_data()->~MethodData();
```

**Review by dholmes-ora — APPROVED** [review]
Thanks Coleen. A couple of further nits on comments but otherwise this seems okay.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.cpp:2675]: Yes, plus releasing the JVMCI thing that MethodData points to also.

File: src/hotspot/share/oops/instanceKlass.cpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/share/oops/instanceKlass.cpp]: The comment means it can't release constant pool CHeap structures.  It should be inclusive of MethodData also.

File: src/hotspot/share/memory/metadataFactory.hpp [hunk 7]
```diff
@@ -61,14 +62,21 @@ class MetadataFactory : AllStatic {
...
       int size = md->size();
-      // Call metadata's deallocate function which will call deallocate fields
+      // Call metadata's deallocate function which will deallocate fields and release_C_heap_structures
       assert(!md->on_stack(), "can't deallocate things on stack");
       assert(!md->is_shared(), "cannot deallocate if in shared spaces");
       md->deallocate_contents(loader_data);
+      // Call the destructor. This is currently used for MethodData which has a member
+      // that needs to be destructed to release resources. Most Metadata derived classes have noop
+      // destructors and/or cleanup using deallocate_contents.
+      // T is a potentially const or volatile qualified pointer. Remove any const
```
Comment (jcking) [inline src/hotspot/share/memory/metadataFactory.hpp:76]: We'll given the template, T by itself is no longer a pointer so its just potentially const or volatile qualified. But that is pedantic.

**Review by jcking — APPROVED** [review]

**Review by dholmes-ora — APPROVED** [review]

**Review by TobiHartmann — APPROVED** [review]
Looks good to me. Thanks for fixing!

File: src/hotspot/share/memory/metadataFactory.hpp (hunk 7 above)
Comment (coleenp) [reply src/hotspot/share/memory/metadataFactory.hpp:76]: I just want to leave it because you're right but pointer is a readable noun here.

**Comment by coleenp** [comment]
Thanks for the reviews David, Justin and Tobias.
/integrate
